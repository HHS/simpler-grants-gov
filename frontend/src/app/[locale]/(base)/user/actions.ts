"use server";

import { z } from "zod";

import { getTranslations } from "next-intl/server";

export type UserProfileValidationErrors = {
  firstName?: string[];
  lastName?: string[];
};

type UserProfileResponse = {
  validationErrors: UserProfileValidationErrors;
};

const validateUserProfileAction = async (formData: FormData) => {
  const t = await getTranslations("UserProfile.validationErrors");
  const schema = z.object({
    firstName: z.string().min(1, { message: t("firstName") }),
    lastName: z.string().min(1, { message: t("lastName") }),
  });

  const validatedFields = schema.safeParse({
    firstName: formData.get("firstName"),
    lastName: formData.get("lastName"),
  });

  // Return early if the form data is invalid (server side validation!)
  if (!validatedFields.success) {
    return validatedFields.error.flatten().fieldErrors;
  }
  return {};
};

export const userProfileAction = async (
  _prevState: UserProfileResponse,
  formData: FormData,
): Promise<UserProfileResponse> => {
  console.log("Received form data to save to user profile", formData);
  const validationErrors = await validateUserProfileAction(formData);
  return {
    validationErrors,
  };
};
