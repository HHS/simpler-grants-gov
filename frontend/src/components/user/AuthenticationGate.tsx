import { getSession } from "src/services/auth/session";

import { getTranslations } from "next-intl/server";
import { ReactNode } from "react";
import { Alert, GridContainer } from "@trussworks/react-uswds";

export async function AuthenticationGate({
  children,
}: {
  children: ReactNode;
}) {
  const session = await getSession();
  const t = await getTranslations("Errors");
  if (!session?.token) {
    return (
      <GridContainer className="margin-top-4">
        <Alert type="error" heading={t("unauthenticated")} headingLevel="h4">
          {t("signInCTA")}
        </Alert>
      </GridContainer>
    );
  }
  return children;
}
