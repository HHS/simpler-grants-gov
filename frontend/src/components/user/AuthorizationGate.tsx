"use server";

import { isEmpty } from "lodash";
import { parseErrorStatus } from "src/errors";
import { AuthorizedDataProvider } from "src/hooks/useAuthorizedData";
import { getSession } from "src/services/auth/session";
import { checkUserPrivilege } from "src/services/fetch/fetchers/userFetcher";
import {
  AuthorizedData,
  FetchedResource,
  FetchedResourceMap,
  ResourcePromiseDefinitions,
} from "src/types/authTypes";
import {
  UserPrivilegeDefinition,
  UserPrivilegeResult,
} from "src/types/userTypes";

import {
  cloneElement,
  JSXElementConstructor,
  PropsWithChildren,
  ReactElement,
  ReactNode,
} from "react";
import { Alert } from "@trussworks/react-uswds";

import { UnauthenticatedMessage } from "./UnauthenticatedMessage";

// requires either `requiredPrivileges` or `resourcePromises` because otherwise why are you using the gate?
type AuthorizationGateProps = {
  onUnauthorized?: (
    children: ReactNode,
    fetchedResources?: FetchedResourceMap,
  ) => ReactNode;
  onUnauthenticated?: () => ReactNode;
  onError?: (e: Error) => ReactNode;
} & (
  | {
      requiredPrivileges: UserPrivilegeDefinition[];
      resourcePromises?: ResourcePromiseDefinitions;
    }
  | {
      resourcePromises: ResourcePromiseDefinitions;
      requiredPrivileges?: UserPrivilegeDefinition[];
    }
  | {
      resourcePromises: ResourcePromiseDefinitions;
      requiredPrivileges: UserPrivilegeDefinition[];
    }
);

// unpack the fetched resource maps to get the first actual error
const findFirstError = (
  fetchedResources: FetchedResourceMap[],
): FetchedResource | undefined => {
  const firstError = fetchedResources.find((resource) =>
    Object.values(resource).some(
      (resourceDefinition) => resourceDefinition.statusCode !== 200,
    ),
  );
  return firstError ? Object.values(firstError)[0] : undefined;
};

// we need to expose status codes and possible errors as well as fetched data
// so that components can use this full context down stream.
// the promises generated here will be consumed in a Promise.all downstream.
// for this to work, will need to make sure any resource promises throw errors
// that include a status code in the cause...
const resolveAndFormatResources = (
  resourcePromises: ResourcePromiseDefinitions,
): Promise<FetchedResourceMap>[] => {
  const fetchResourcePromises = Object.entries(resourcePromises).map(
    ([resourceName, resourcePromise]) => {
      return resourcePromise
        .then((resourceData) => {
          return {
            [resourceName]: {
              data: resourceData,
              statusCode: 200,
            } as FetchedResource,
          };
        })
        .catch((e: Error) => {
          return {
            [resourceName]: {
              error: e.message,
              statusCode: parseErrorStatus(e),
            } as FetchedResource,
          };
        });
    },
  );
  return fetchResourcePromises;
};

// calls the API to check all required privileges,
// and formats results into a format to be used downstream
const checkRequiredPrivileges = async (
  token: string,
  userId: string,
  privileges: UserPrivilegeDefinition[],
): Promise<UserPrivilegeResult[]> => {
  const privilegeCheckResults = await Promise.all(
    privileges.map((privilege) => {
      return checkUserPrivilege(token, userId, privilege)
        .then(() => {
          return { ...privilege, authorized: true };
        })
        .catch((e: Error) => {
          if (parseErrorStatus(e) === 403) {
            return { ...privilege, authorized: false };
          }
          throw e;
        });
    }),
  );
  return privilegeCheckResults;
};

/*
  makes calls for required resources, if passed
  checks for required permissions, if passed
  optionally returns results of onUnauthorized function, if user is unauthorized based on calls to required resources
  passes results from required resource and required permission calls via both props and context
*/
export async function AuthorizationGate({
  children,
  onUnauthorized,
  onUnauthenticated = () => <UnauthenticatedMessage />,
  onError = (e: Error) => (
    <Alert heading={e.message} type="error" headingLevel="h2" />
  ),
  requiredPrivileges,
  resourcePromises,
}: PropsWithChildren<AuthorizationGateProps>) {
  let userPrivileges: UserPrivilegeResult[] = [];
  let allResources = {};

  const session = await getSession();

  if (!session?.token) {
    return onUnauthenticated();
  }

  // check privileges
  if (requiredPrivileges) {
    userPrivileges = await checkRequiredPrivileges(
      session.token,
      session.user_id,
      requiredPrivileges,
    );
  }

  const mappedResourcePromises =
    resourcePromises && !isEmpty(resourcePromises)
      ? resolveAndFormatResources(resourcePromises)
      : undefined;

  // fetch resources and check for 403s
  if (mappedResourcePromises) {
    try {
      // Note: there's a potential performance gain here if we make these fetches in parallel with the user privileges calls
      const fetchedResources = await Promise.all(mappedResourcePromises);
      allResources = fetchedResources.reduce(
        (all, resource) => ({ ...all, ...resource }),
        {},
      );
      // Note: we will handle errors based on the order that their respective promise was passed in. If multiple promises throw errors
      // any after the first error encountered will be swallowed, as we have not accounted for the ability to handle multiple errors.
      // We can improve this later
      const firstError = findFirstError(fetchedResources);
      if (firstError) {
        if (firstError.statusCode === 403 && onUnauthorized) {
          return onUnauthorized(children, allResources);
        }
        if (firstError.statusCode !== 403) {
          return onError(new Error(firstError.error));
        }
      }
    } catch (e) {
      return onError(e as Error);
    }
  }

  const authorizedData: AuthorizedData = {
    fetchedResources: allResources,
    confirmedPrivileges: userPrivileges,
  };

  // Allows for prop drilling of fetched resources into immediate child server components, as
  // usage of the client side context provider won't be available to server components.
  // Assumes any immediate children of the gate that want to accept fetchedResources via props
  // have an optional `fetchedResources` prop in place that can accept the added data here
  // Note: this does NOT work if the gate is used in a layout component - if you're going to prop
  // drill make sure you do it from a page component or further down the render chain
  const childrenWithResources = children
    ? cloneElement(
        children as ReactElement<
          { authorizedData?: object },
          string | JSXElementConstructor<unknown>
        >,
        { authorizedData },
      )
    : children;

  // AuthorizedDataProvider allows any client component children of the gate to receive
  // fetched resources via context.
  return (
    <AuthorizedDataProvider value={authorizedData}>
      {childrenWithResources}
    </AuthorizedDataProvider>
  );
}
