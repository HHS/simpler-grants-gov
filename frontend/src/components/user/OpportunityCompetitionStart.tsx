"use client";

import { useFeatureFlags } from "src/hooks/useFeatureFlags";
import { Competition } from "src/types/competitionsResponseTypes";

import StartApplicationModal from "src/components/workspace/StartApplicationModal";

export const OpportunityCompetitionStart = ({
  competitions,
  opportunityTitle,
}: {
  competitions: [Competition];
  opportunityTitle: string;
}) => {
  const { checkFeatureFlag } = useFeatureFlags();
  const openCompetitions = competitions.filter(({ is_open }) => is_open);

  if (!openCompetitions.length || checkFeatureFlag("applyFormPrototypeOff")) {
    return <></>;
  } else {
    return (
      <>
        <StartApplicationModal
          opportunityTitle={opportunityTitle}
          competitionId={openCompetitions[0].competition_id}
        />
      </>
    );
  }
};
