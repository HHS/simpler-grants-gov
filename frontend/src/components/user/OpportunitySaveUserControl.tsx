"use client";

import { useFeatureFlags } from "src/hooks/useFeatureFlags";
import { useUser } from "src/services/auth/useUser";

import { useTranslations } from "next-intl";
import Link from "next/link";
import { useParams } from "next/navigation";
import { useEffect, useRef, useState } from "react";
import { ModalRef, ModalToggleButton } from "@trussworks/react-uswds";

import { LoginModal } from "src/components/LoginModal";
import SaveButton from "src/components/SaveButton";
import { USWDSIcon } from "src/components/USWDSIcon";

const SAVED_OPPS_PAGE_LINK = "/saved-grants";

export const OpportunitySaveUserControl = () => {
  const t = useTranslations("OpportunityListing");
  const modalRef = useRef<ModalRef>(null);
  const params = useParams();
  const opportunityId = String(params.id);

  const { user } = useUser();
  const [saved, setSaved] = useState(false);
  const [showMessage, setshowMessage] = useState(false);
  const [savedError, setSavedError] = useState(false);
  const [loading, setLoading] = useState(false);

  const closeMessage = () => {
    setshowMessage(false);
  };

  const userSavedOppCallback = async () => {
    setLoading(true);

    const method = saved ? "DELETE" : "POST";
    try {
      const res = await fetch("/api/user/saved-opportunities", {
        method,
        headers: {
          opportunityId,
        },
      });
      if (res.ok && res.status === 200) {
        const data = (await res.json()) as { type: string };
        data.type === "save" ? setSaved(true) : setSaved(false);
      } else {
        setSavedError(true);
      }
    } catch (error) {
      setSavedError(true);
      console.error(error);
    } finally {
      setshowMessage(true);
      setLoading(false);
    }
  };

  useEffect(() => {
    if (!user?.token) return;
    setLoading(true);
    fetch(`/api/user/saved-opportunities/${opportunityId}`)
      .then((res) => (res.ok && res.status === 200 ? res.json() : null))
      .then((data) => {
        data && setSaved(true);
      })
      .finally(() => {
        setLoading(false);
      })
      .catch((error) => {
        console.error(error);
      });
  }, [opportunityId, user]);

  const messageText = saved
    ? savedError
      ? t("save_message.error_unsave")
      : t.rich("save_message.save", {
          linkSavedGrants: (chunks) => (
            <Link className="text-black" href={SAVED_OPPS_PAGE_LINK}>
              {chunks}
            </Link>
          ),
        })
    : savedError
      ? t("save_message.error_save")
      : t("save_message.unsave");

  const { checkFeatureFlag } = useFeatureFlags();
  if (!checkFeatureFlag("savedOpportunitiesOn")) return null;
  return (
    <>
      {user?.token ? (
        <SaveButton
          buttonClick={userSavedOppCallback}
          messageClick={closeMessage}
          buttonId="opp-save-button"
          defaultText={t("save_button.save")}
          error={savedError}
          messageText={messageText}
          message={showMessage}
          loading={loading}
          loadingText={t("save_button.loading")}
          saved={saved}
          savedText={t("save_button.saved")}
        />
      ) : (
        <>
          <ModalToggleButton
            modalRef={modalRef}
            opener
            className="simpler-save-button usa-button usa-button--outline"
          >
            <USWDSIcon name="star_outline" />
            {t("save_button.save")}
          </ModalToggleButton>
          <LoginModal
            modalRef={modalRef as React.RefObject<ModalRef>}
            helpText={t("save_login_modal.help")}
            buttonText={t("save_login_modal.button")}
            closeText={t("save_login_modal.close")}
            descriptionText={t("save_login_modal.description")}
            titleText={t("save_login_modal.title")}
            modalId="opp-save-login-modal"
          />
        </>
      )}
    </>
  );
};
