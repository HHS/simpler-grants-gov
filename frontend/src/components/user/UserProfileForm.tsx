"use client";

import {
  userProfileAction,
  UserProfileValidationErrors,
} from "src/app/[locale]/(base)/user/actions";
import { UserDetail } from "src/types/userTypes";

import { useTranslations } from "next-intl";
import { useActionState } from "react";
import { Button, Label, TextInput } from "@trussworks/react-uswds";

import { ConditionalFormActionError } from "src/components/ConditionalFormActionError";
import { RequiredFieldIndicator } from "src/components/RequiredFieldIndicator";

const UserProfileValidationError =
  ConditionalFormActionError<UserProfileValidationErrors>;

export function UserProfileForm({ userDetails }: { userDetails: UserDetail }) {
  const t = useTranslations("UserProfile");

  const [state, formAction] = useActionState(userProfileAction, {
    validationErrors: {},
  });

  return (
    <form action={formAction}>
      <Label htmlFor="firstName">
        <span>{t("inputs.firstName")}</span>
        <RequiredFieldIndicator> *</RequiredFieldIndicator>
      </Label>
      <UserProfileValidationError
        fieldName="firstName"
        errors={state.validationErrors}
      />
      <TextInput
        id="edit-user-first-name"
        name="firstName"
        type="text"
        value={state.first_name || userDetails.first_name}
      />
      <Label htmlFor="middle-name">{t("inputs.middleName")}</Label>
      <TextInput
        id="edit-user-middle-name"
        name="middleName"
        type="text"
        value={state.middle_name || userDetails.middle_name}
      />
      <Label htmlFor="lastName">
        <span>{t("inputs.lastName")}</span>
        <RequiredFieldIndicator> *</RequiredFieldIndicator>
      </Label>
      <UserProfileValidationError
        fieldName="lastName"
        errors={state.validationErrors}
      />
      <TextInput
        id="edit-user-last-name"
        name="lastName"
        type="text"
        value={state.last_name || userDetails.last_name}
      />
      <Label htmlFor="email">{t("inputs.email")}</Label>
      <TextInput
        id="edit-user-email"
        name="email"
        type="text"
        defaultValue={userDetails.email}
        disabled
      />
      <Button type="submit" className="margin-top-4">
        {t("save")}
      </Button>
    </form>
  );
}
