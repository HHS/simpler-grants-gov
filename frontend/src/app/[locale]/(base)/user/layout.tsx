import { Metadata } from "next";
import { LayoutProps } from "src/types/generalTypes";
import { LocalizedPageProps } from "src/types/intl";

import { getTranslations } from "next-intl/server";

import { AuthenticationGate } from "src/components/user/AuthenticationGate";

export async function generateMetadata({
  params,
}: LocalizedPageProps): Promise<Metadata> {
  const { locale } = await params;
  const t = await getTranslations({ locale });
  const meta: Metadata = {
    title: t("UserAccount.pageTitle"),
    description: t("Index.metaDescription"),
  };
  return meta;
}
export default function UserLayout({ children }: LayoutProps) {
  return <AuthenticationGate>{children}</AuthenticationGate>;
}
