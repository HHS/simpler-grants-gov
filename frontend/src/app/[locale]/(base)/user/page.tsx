import { getSession } from "src/services/auth/session";

import { getTranslations } from "next-intl/server";
import { GridContainer } from "@trussworks/react-uswds";

import { UserProfileForm } from "src/components/user/UserProfileForm";

export default async function UserProfile() {
  const t = await getTranslations("UserProfile");

  const session = await getSession();
  if (!session?.email) {
    // this won't happen, as email is required on sessions, and we're wrapping this in an auth gate in the layout
    console.error("no user session, or user has no email address");
    return;
  }

  // fetch name info from user endpoint
  // gate on feature flag
  return (
    <GridContainer>
      <h1>{t("title")}</h1>
      <UserProfileForm email={session.email} />
    </GridContainer>
  );
}
