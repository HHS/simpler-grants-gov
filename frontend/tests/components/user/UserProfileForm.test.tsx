import { render, screen } from "@testing-library/react";
import { noop } from "lodash";
import { fakeUser } from "src/utils/testing/fixtures";
import { useTranslationsMock } from "src/utils/testing/intlMocks";

import { UserProfileForm } from "src/components/user/UserProfileForm";

const mockUseActionState = jest.fn();

jest.mock("react", () => ({
  ...jest.requireActual<typeof import("react")>("react"),
  useActionState: () => mockUseActionState() as unknown,
}));

jest.mock("src/app/[locale]/(base)/user/account/actions", () => ({
  userProfileAction: noop,
}));

jest.mock("next-intl", () => ({
  useTranslations: () => useTranslationsMock(),
}));

describe("userProfileForm", () => {
  afterEach(() => {
    jest.resetAllMocks();
  });
  it("matches snapshot", () => {
    mockUseActionState.mockReturnValue([{}, noop, false]);
    const { container } = render(<UserProfileForm userDetails={fakeUser} />);
    expect(container).toMatchSnapshot();
  });
  it("renders validation errors", () => {
    mockUseActionState.mockReturnValue([
      { validationErrors: { firstName: ["some error"] } },
      noop,
      false,
    ]);

    render(<UserProfileForm userDetails={fakeUser} />);

    const alert = screen.getByRole("alert");

    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent("some error");
  });
  it("displays values from userDetail props if form is not submitted", () => {
    if (!fakeUser.profile) {
      throw new Error("this will not happen, just here to get TS to behave");
    }
    mockUseActionState.mockReturnValue([{}, noop, false]);

    render(<UserProfileForm userDetails={fakeUser} />);

    expect(
      screen.getByDisplayValue(fakeUser.profile.first_name),
    ).toBeInTheDocument();
    expect(
      screen.getByDisplayValue(fakeUser.profile.last_name),
    ).toBeInTheDocument();
  });
  it("displays values from form action when available", () => {
    mockUseActionState.mockReturnValue([
      {
        data: {
          profile: {
            first_name: "something",
            middle_name: "else",
            last_name: "entirely",
          },
        },
      },
      noop,
      false,
    ]);

    render(<UserProfileForm userDetails={fakeUser} />);

    expect(screen.getByDisplayValue("something")).toBeInTheDocument();
    expect(screen.getByDisplayValue("else")).toBeInTheDocument();
    expect(screen.getByDisplayValue("entirely")).toBeInTheDocument();
  });
  it("displays error message on error", () => {
    mockUseActionState.mockReturnValue([
      {
        errorMessage: "big error",
      },
      noop,
      false,
    ]);

    render(<UserProfileForm userDetails={fakeUser} />);

    const alert = screen.getByRole("heading");
    expect(alert).toBeInTheDocument();
    expect(alert).toHaveTextContent("errorHeading");
  });
});
