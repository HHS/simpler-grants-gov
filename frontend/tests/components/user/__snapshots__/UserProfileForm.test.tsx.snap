// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`userProfileForm matches snapshot 1`] = `
<div>
  <form
    action="javascript:throw new Error('A React form was unexpectedly submitted. If you called form.submit() manually, consider using form.requestSubmit() instead. If you\\'re trying to use event.stopPropagation() in a submit event handler, consider also calling event.preventDefault().')"
  >
    <label
      class="usa-label"
      data-testid="label"
      for="firstName"
    >
      <span>
        inputs.firstName
      </span>
      <span
        class="usa-hint usa-hint--required"
        data-testid="required-field-indicator"
      >
         *
      </span>
    </label>
    <input
      class="usa-input"
      data-testid="textInput"
      id="edit-user-first-name"
      name="firstName"
      type="text"
      value="joe"
    />
    <label
      class="usa-label"
      data-testid="label"
      for="middle-name"
    >
      inputs.middleName
    </label>
    <input
      class="usa-input"
      data-testid="textInput"
      id="edit-user-middle-name"
      name="middleName"
      type="text"
    />
    <label
      class="usa-label"
      data-testid="label"
      for="lastName"
    >
      <span>
        inputs.lastName
      </span>
      <span
        class="usa-hint usa-hint--required"
        data-testid="required-field-indicator"
      >
         *
      </span>
    </label>
    <input
      class="usa-input"
      data-testid="textInput"
      id="edit-user-last-name"
      name="lastName"
      type="text"
      value="quisling"
    />
    <label
      class="usa-label"
      data-testid="label"
      for="email"
    >
      inputs.email
    </label>
    <input
      class="usa-input"
      data-testid="textInput"
      disabled=""
      id="edit-user-email"
      name="email"
      type="text"
      value="not-the-real-email@fake.com"
    />
    <button
      class="usa-button margin-top-4"
      data-testid="button"
      type="submit"
    >
      save
    </button>
  </form>
</div>
`;
